### DISCUSSION TITLE
Dynamic achievement sorting on game page

### DISCUSSION THREAD

**Discussion by Krylan** [RetroAchievements/RAWeb#1226] [Ideas]
I have some idea for changing the way how achievements are sorted. Before I go into introducing such changes, I want to consult my proposition with you, so maybe there will be some more ideas on how to make it better.

Right now on game page, user can select one of the four pre-defined sorting rules for the achievement list below. By clicking a link, sorting gets applied in ascending manner. If user wants to get descending order, then they need to click the link again.
There are some disadvantages of the way, how it's actually working:
– every time user changes way of sorting, whole page gets reloaded,
– if user wants to change ordering, they need to click the link twice (so page gets reloaded twice as well)
– search engine bots visits those links as well, and they perceive those addressess to be different, but the content of the page is exactly the same (only order of items changes),

My proposition to improve this element, is to change links with parametrized URLs to selects. After changing the option, instead of changing URL, Javascript should reorder achievements list (we don't have anything else to do, as all achievements are loaded already on the page). Also, it should take less space, letting us to provide even more functionality. In my proposition, I added a display setting, wchich lets an user to hide unlocked or locked achievements and see the other ones.

Here is a small visualization, what I have on my mind:
BEFORE
![image](https://user-images.githubusercontent.com/26091129/201394369-7d50bef9-f030-4085-b064-222b7403a96b.png)
AFTER
![image](https://user-images.githubusercontent.com/26091129/201394397-bae60f08-3956-406a-a12a-2caaa4a27609.png)
(it's not ideal in terms of spacing and alignment, but I was using a new design app I'm not used to, and it even crashed during work, so I just made a screenshot of my actual work. It's just for preview)
Alternatively, we could join sorting and ordering selects into one, and just list all possible options with ordering included.

Let me know, if you support this idea, or if you have some additional ideas on how to make it better.

**Comment by Jamiras** [comment]
I never use those sort fields, and I'm not really sure why anyone would, so I'm not sure adding more options is worth the effort.

The one thing I would sometimes like to do is to see the locked/unlocked achievements in their original order (without grouping the unlocked first), as the display order often provides insight as to the order of achievements so I can figure out whereabouts in the game something I missed occurs.

I see this being more important once tags are implemented, as the user would likely want to filter on specific tags (i.e. missables).

**Reply by Krylan** [reply]
@Jamiras while I don't see the reason behind sorting on title (alphabetically), I see "Won by" and "Points" sorting quite usable for fast look on most difficult/absorbing achievements.
Yesterday, I made a community poll on Discord to get more insight from the users:
:bar_chart: On RA website, do you use achievement sorting function on game pages?
A) Yeah, I use it regularly (9 votes - 26%)
B) I use it from time to time (20 votes - 57%)
C) I don't use it at all (6 votes - 17%)
Total votes: 35

It seems that most of voters are in fact using the sorting function. There was an opinion, that sorting by unlock count helps in finding the best course of action to get achievements. There was also an idea to add sorting by retro points (aka "white points") – right now the alternative for that is sorting by unlock count and find the most valued achievement among the less unlocked achievements.

I was also thinking about filtering the list according to upcoming tags system. There could be something like "advanced" dropdown, or just a box with tags below, which could be checked/unchecked individually to show & hide achievements. In fact, we could already filter missables based on "[m]" text in achievement title, but I would wait with it until proper tags will come out.

Anyway, I would like to provide this improvement for the users (it would also relieve server a bit by excluding unecessary requests for pages with different sorting), but also for bots and crawlers, so they won't look after sorting URLs for each game (this would be like thousands of URLs).

**Comment by Jamiras** [comment]
Also, related: #764 (ability to hide unlocked achievements [hardcore/softcore individually])

**Comment by Jamiras** [comment]
Something else to consider: Both the standalone overlay and RetroArch list unearned achievements before earned achievements. The website displays earned first.

**Comment by DouglasKGlover** [comment]
Bumping this to say I'd highly value better sort options, yet even the current sorting options being changed to client-side JS would be a huge improvement both for users (quicker) and the site itself (reduced server load with all the page reloads from sorting).

My personal use case, and I do this regularly, is to check the "Won by" sort to see what the hardest few achievements are (as they've been earned by the fewest people) to gauge whether I'd be able to Master the set.
